### PR TITLE
New linker intrinsics

### PR DESCRIPTION
**Pull request by tlakollo**
Add new intrinsics for:
- GetConstructors
- GetEvents
- GetFields
- GetProperties
- GetMethods
- GetNestedTypes
- GetMember
- GetMembers
Add tests

This should fix the problem in #1724

### CODE REVIEW

File: test/Mono.Linker.Tests.Cases/Reflection/ConstructorsUsedViaReflection.cs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		[RecognizedReflectionAccessPattern]
+		[Kept]
+		static void TestWithUnknownBindingFlags (BindingFlags bindingFlags)
+		{
+			// Since the binding flags are not known linker should mark all constructors on the type
+			var constructors = typeof (UnknownBindingFlags).GetConstructors (bindingFlags);
+		}
+
+		[Kept]
+		static void TestNullType ()
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/ConstructorsUsedViaReflection.cs:47]: RecognizedReflectionAccessPattern

File: test/Mono.Linker.Tests.Cases/Reflection/EventUsedViaReflection.cs [hunk 2]
```diff
@@ -144,8 +144,8 @@ static void TestIfElse (int i)
 			typeof (BaseClass), nameof (BaseClass.PublicEventOnBase), (Type[]) null)]
 		static void TestEventInBaseType ()
 		{
-			typeof (DerivedClass).GetEvent ("ProtectedEventOnBase"); // Will not mark anything as it only works on public events
-			typeof (DerivedClass).GetEvent ("PublicEventOnBase");
+			var eventInfo = typeof (DerivedClass).GetEvent ("ProtectedEventOnBase"); // Will not mark anything as it only works on public events
+			eventInfo = typeof (DerivedClass).GetEvent ("PublicEventOnBase");
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/EventUsedViaReflection.cs:148]: What's the purpose of this change (also below) - I would remove it.

File: test/Mono.Linker.Tests.Cases/Reflection/EventsUsedViaReflection.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		[Kept]
+		[RecognizedReflectionAccessPattern]
+		static void TestGetEvents ()
+		{
+			var events = typeof (Foo).GetEvents ();
+		}
+
+		[Kept]
+		// The event will not be kept as it's internal and the behavior of Type.GetEvent(string) is to only return public events
+		// But we don't mark it as unrecognized access pattern - we did recognize it fully, just didn't find the event being asked for
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/EventsUsedViaReflection.cs:39]:
```suggestion
		// The event will not be kept as it's internal and the behavior of Type.GetEvents() is to only return public events
		// But we don't mark it as unrecognized access pattern - we did recognize it fully, just didn't find the event being asked for
```

File: test/Mono.Linker.Tests.Cases/Reflection/EventsUsedViaReflection.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+		// The behavior of the code will not change by linking it:
+		//   - Without linking the GetEvent will return null
+		//   - After linking the GetEvent will still return null
+		// We also don't mark it as recognized pattern since we didn't mark anything
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/EventsUsedViaReflection.cs:43]:
```suggestion
		//   - Without linking the GetEvents will return null
		//   - After linking the GetEvents will still return null
		// We also don't mark it as recognized pattern since we didn't mark anything
```

File: test/Mono.Linker.Tests.Cases/Reflection/EventsUsedViaReflection.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+				myType = typeof (IfClass);
+			} else {
+				myType = typeof (ElseClass);
+			}
+			String myString;
+			if (i == 1) {
+				myString = "IfEvent";
+			} else {
+				myString = "ElseEvent";
+			}
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/EventsUsedViaReflection.cs:106]: This is not used for anything

File: test/Mono.Linker.Tests.Cases/Reflection/FieldsUsedViaReflection.cs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+				myType = typeof (IfClass);
+			} else {
+				myType = typeof (ElseClass);
+			}
+			String myString;
+			if (i == 1) {
+				myString = "ifField";
+			} else {
+				myString = "elseField";
+			}
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/FieldsUsedViaReflection.cs:93]: Not used for anything

File: test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+
+namespace Mono.Linker.Tests.Cases.Reflection
+{
+	[SetupCSharpCompilerToUse ("csc")]
+	public class MemberUsedViaReflection
+	{
+		public static void Main ()
+		{
+			// Normally calls to GetMember use regex to find values, we took a conservative approach
+			// and preserve not based on the string passed but on the binding flags requirements
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs:15]: I don't get this - regex?
I'm fine if we don't implement the name based lookups (this is not very common), but the comment here is weird.

Comment (vitek-karas) [reply test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs:15]: I see - well I would not call "C*" a regex (it's actually not a valid regex) - more like "wildcard".
Can you please rename anything with "Regex" in this class - it doesn't matter to what really, but regex is confusing.

File: test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+			var members = typeof (UnknownBindingFlags).GetMember ("*FoundViaRegex", bindingFlags);
+		}
+
+		[RecognizedReflectionAccessPattern]
+		[Kept]
+		static void TestWithMemberTypes ()
+		{
+			// Here we took the same conservative approach, instead of understanding MemberTypes we only use
+			// the information in the binding flags requirements and keep all the MemberTypes
+			var members = typeof (TestMemberTypes).GetMember ("*FoundViaRegex", MemberTypes.Method, BindingFlags.Public);
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs:63]: BTW: "*C" will not work if I read the code in CoreLib correctly - it can only handle prefixes, not suffixes.

File: test/Mono.Linker.Tests.Cases/Reflection/MembersUsedViaReflection.cs [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+		static void TestWithBindingFlags ()
+		{
+			var members = typeof (MembersBindingFlags).GetMembers (BindingFlags.Public);
+		}
+
+		[RecognizedReflectionAccessPattern]
+		[Kept]
+		static void TestWithUnknownBindingFlags (BindingFlags bindingFlags)
+		{
+			// Since the binding flags are not known linker should mark all constructors on the type
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/Reflection/MembersUsedViaReflection.cs:42]:
```suggestion
			// Since the binding flags are not known linker should mark all members on the type
```

**Comment by vitek-karas** [comment]
Do we have a tracking work item to make corresponding changes to the runtime repo?
* Change `Type.GetMembers` annotation to not use `All` and instead list all of the other bits (the same way linker treats it as an intrinsic)
* Remove suppressions in reflection for the warning this resolves - for example `Type.GetProperties()` has a suppression which should not be necessary after this change. Might be worth going over all potential candidates (the less suppressions the better).
* Remove suppressions in the rest of libraries - @eerhardt is probably going to have an idea where this was used so far.

File: test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs (hunk 8 above)
Comment (tlakollo) [reply test/Mono.Linker.Tests.Cases/Reflection/MemberUsedViaReflection.cs:63]: Yup, thanks for the heads up 
https://github.com/dotnet/runtime/blob/0026a202ab4087d237f6be8824dff322d3a7d6be/src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs#L2059

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 10]
```diff
@@ -1116,6 +1181,106 @@ public override bool HandleCall (MethodBody callingMethodBody, MethodReference c
...
+
+
+				//
+				// GetMember (String)
+				// GetMember (String, BindingFlags)
+				// GetMember (String, MemberTypes, BindingFlags)
+				//
+				case IntrinsicId.Type_GetMember when calledMethod.DeclaringType.Namespace == "System"
+					&& calledMethod.DeclaringType.Name == "Type"
+					&& calledMethod.HasParameterOfType (0, "System", "String"): {
```
Comment (vitek-karas) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:1249]: This doesn't make sense to me - we check exactly this in the first switch - so no reason to check the namespace, typename and first parameter here again. My comment was about the fact that the code below checks that BindingFlags parameter is in a certain position based on the number of parameters to the method. And if those checks don't succeed we fail. But the switch above which detects the `IntrinsicId` does not perform those checks.

So for example, if there was a method Type.GetMember(String, MemberTypes)  (no binding flags), the first switch would accept it as recognized and detected it as `IntrinsicId.Type_GetMember`, it would get here, and the code below would fail because the binding flags checks would fail - it would end up effectively crashing the linker. That's not good. The check which detects the `IntrinsicId` needs to be very precise, so that the code in here (HandleCall) can't ever fail (well at least for reasons of not recognizing the method).

**Review by vitek-karas — APPROVED** [review]
Other than the nit in https://github.com/mono/linker/pull/1829#discussion_r576207947 this looks good.

**Comment by tlakollo** [comment]
The lint command updated the dotnet-format version from version '4.1.131201' to 'version '5.0.211103' which seems to be causing this error, I don't see any problems with the file, and the lint in my workstation is still 4.1.131201 so it won't fix the whitespace errors

**Comment by marek-safar** [comment]
@tlakollo you need to merge with master locale to repro that
